src/languages/reasonml.js

```diff
@@ -6,301 +6,63 @@ Author: Gidi Meir Morris <oss@gidi.io>
-        relevance: 0,
-        contains: PARAMS_CONTENTS
-      }
-    ]
-  };
-
-  const FUNCTION_BLOCK_MODE = {
-    className: 'function',
-    relevance: 0,
-    keywords: KEYWORDS,
-    variants: [
-      {
-        begin: '\\s(\\(\\.?.*?\\)|' + RE_IDENT + ')\\s*=>',
-        end: '\\s*=>',
-        returnBegin: true,
-        relevance: 0,
-        contains: [
-          {
-            className: 'params',
-            variants: [
-              { begin: RE_IDENT },
-              { begin: RE_PARAM },
-              { begin: /\(\s*\)/ }
-            ]
-          }
-        ]
-      },
-      {
-        begin: '\\s\\(\\.?[^;\\|]*\\)\\s*=>',
-        end: '\\s=>',
-        returnBegin: true,
-        relevance: 0,
-        contains: [
-          {
-            className: 'params',
-            relevance: 0,
-            variants: [ PARAMS_MODE ]
-          }
-        ]
-      },
-      { begin: '\\(\\.\\s' + RE_IDENT + '\\)\\s*=>' }
-    ]
-  };
-  MODULE_ACCESS_CONTENTS.push(FUNCTION_BLOCK_MODE);
-
-  const CONSTRUCTOR_MODE = {
-    className: 'constructor',
-    begin: RE_MODULE_IDENT + '\\(',
-    end: '\\)',
-    illegal: '\\n',
-    keywords: KEYWORDS,
-    contains: [
-      hljs.QUOTE_STRING_MODE,
-      OPERATOR_MODE,
-      {
-        className: 'params',
-        begin: '\\b' + RE_IDENT
-      }
-    ]
-  };
-
-  const PATTERN_MATCH_BLOCK_MODE = {
-    className: 'pattern-match',
-    begin: '\\|',
-    returnBegin: true,
-    keywords: KEYWORDS,
-    end: '=>',
-    relevance: 0,
-    contains: [
-      CONSTRUCTOR_MODE,
-      OPERATOR_MODE,
-      {
-        relevance: 0,
-        className: 'constructor',
-        begin: RE_MODULE_IDENT
-      }
-    ]
-  };
-
-  const MODULE_ACCESS_MODE = {
-    className: 'module-access',
-    keywords: KEYWORDS,
-    returnBegin: true,
-    variants: [
-      { begin: "\\b(" + RE_MODULE_IDENT + "\\.)+" + RE_IDENT },
-      {
-        begin: "\\b(" + RE_MODULE_IDENT + "\\.)+\\(",
-        end: "\\)",
-        returnBegin: true,
-        contains: [
-          FUNCTION_BLOCK_MODE,
-          {
-            begin: '\\(',
-            end: '\\)',
-            relevance: 0,
-            skip: true
-          }
-        ].concat(MODULE_ACCESS_CONTENTS)
-      },
-      {
-        begin: "\\b(" + RE_MODULE_IDENT + "\\.)+\\{",
-        end: /\}/
-      }
-    ],
-    contains: MODULE_ACCESS_CONTENTS
-  };
-
-  PARAMS_CONTENTS.push(MODULE_ACCESS_MODE);
-
   return {
     name: 'ReasonML',
     aliases: [ 're' ],
-    keywords: KEYWORDS,
+    keywords: {
+      $pattern: '[a-z_]\\w*!?',
+      keyword:
+        'and as asr assert begin class constraint do done downto else end esfun '
+        + 'exception external for fun function functor if in include inherit '
+        + 'initializer land lazy let lor lsl lsr lxor mod module mutable new '
+        + 'nonrec object of open or pri pub rec sig struct switch then to try '
+        + 'type val virtual when while with',
+      built_in:
```

Comment:
These look like `type` to me.

Reply:
These are built-in types. I would like to keep compatibility with OCaml syntax, which also defines them as built-in:

https://github.com/highlightjs/highlight.js/blob/73a398ddca4339273fd1203f1ecabab86d75de5c/src/languages/ocaml.js#L26-L27

There are also other languages that do something similar, like Dart:

https://github.com/highlightjs/highlight.js/blob/73a398ddca4339273fd1203f1ecabab86d75de5c/src/languages/dart.js#L197-L198

I made the definition more explicit in https://github.com/highlightjs/highlight.js/pull/3795/commits/d1c3e5833ab7ea627cfdfa0fb275813e69540dcd. Wdyt?

Reply:
>  to keep compatibility with OCaml syntax,

Perhaps Ocaml should also change?  What is it that we do consider a type in ReasonML/OCaml?

Reply:
A type in ReasonML/OCaml is an abstraction that represents a set of values, same as in any other typed language. There are [built-in types](https://ocaml.org/docs/data-types#built-in-compound-types) and [user-defined types](https://ocaml.org/docs/data-types#a-simple-custom-type).

Reply:
I think it is useful to store built-in types in a separate scope, for the sake of having higher fidelity precisely. Note this example:

<img width="242" alt="imagen" src="https://github.com/highlightjs/highlight.js/assets/220424/31134fac-0ddf-4871-a827-9b0a0960259e">

Reply:
Do other highlighters consider these to be separate concepts worthy of separate highlighiting?

Reply:
The vscode-ocaml grammar has a separate highlight for built-in types:

https://github.com/ocamllabs/vscode-ocaml-platform/blob/a170e84a20078a816184ae705b7ee4b6896c4a20/syntaxes/ocaml.json#L617-L619